### PR TITLE
Worksheet_D tasks

### PR DESCRIPTION
**Pull request by Phil-Sparkes**

### CODE REVIEW

File: worksheet_D/oxo.py [hunk 1]
```diff
@@ -1,26 +1,50 @@
...
+            if self.gameboard[squares] == 0:
+                return False
+
+        return True
 
     def get_winner(self):
         """ If a player has three in a row, return 1 or 2 depending on which player.
             Otherwise, return 0. """
-        raise NotImplementedError("TODO: implement get_winner")
+        for x in xrange(3):
```
Comment (edpowley) [inline worksheet_D/oxo.py:33]: You appear to be using `x` as both an x coordinate and a y coordinate here, which is confusing.

File: worksheet_D/oxo.py [hunk 2]
```diff
@@ -1,26 +1,50 @@
...
+
+            # Checks for horizontal and vertical lines of 3
+            if self.gameboard[x] == self.gameboard[x+3] and self.gameboard[x+3] == self.gameboard[x+6]:
```
Comment (edpowley) [inline worksheet_D/oxo.py:36]: This needs a comment to explain the maths. Or maybe you could use `self.get_square` to make it more self-explanatory.

Note also that Python lets you shorten this:
`if a == b and b == c:`
to this:
`if a == b == c:`

File: worksheet_D/oxo.py [hunk 3]
```diff
@@ -1,26 +1,50 @@
...
+                return self.gameboard[x]
+            if self.gameboard[x*3] == self.gameboard[x*3+1] and self.gameboard[x*3+1] == self.gameboard[x*3+2]:
+                return self.gameboard[x*3]
+
+        # Checks for diagonal lines of 3
+        if self.gameboard[0] == self.gameboard[4] and self.gameboard[4] == self.gameboard[8]:
```
Comment (edpowley) [inline worksheet_D/oxo.py:42]: The "magic numbers" 0,4,8 and 2,4,6 need explaining

File: worksheet_D/oxo.py [hunk 4]
```diff
@@ -72,9 +96,12 @@ def input_square():
             # Move was played successfully, so check for a winner
             winner = board.get_winner()
             if winner != 0:
+                board.show()
```
Comment (edpowley) [inline worksheet_D/oxo.py:99]: Yep, I probably should have added this to the template 😄

File: worksheet_D/oxo.py [hunk 5]
```diff
@@ -1,26 +1,50 @@
 class OxoBoard:
     def __init__(self):
         """ The initialiser. Initialise any fields you need here. """
-        raise NotImplementedError("TODO: implement __init__")
+        self.gameboard = [0, 0, 0, 0, 0, 0, 0, 0, 0]
```
Comment (edpowley) [inline worksheet_D/oxo.py:4]: There are neater ways to initialise the list, e.g. with a loop or using list multiplication

**Review by edpowley — COMMENTED** [review]
Code looks good, but would benefit from more comments. Other than that, only a few minor issues.

If you want to address these issues before the deadline on Monday, feel free to do so and open a new pull request. Otherwise I will take this version as your submission.

File: worksheet_D/oxo.py (hunk 2 above)
Comment (edpowley) [reply worksheet_D/oxo.py:36]: Just noticed another problem with this: what happens if there's a row of three empty squares?

Comment (Phil-Sparkes) [reply worksheet_D/oxo.py:36]: A row of empty squares will return a value of 0 which is the return for no winner

Comment (Phil-Sparkes) [reply worksheet_D/oxo.py:36]: After testing this it produced a bug so has been changed now

**Pull request closed without merging**